system/jlib/jsort.cpp

```diff
@@ -703,6 +781,212 @@ void msortvecstableinplace(void ** rows, size32_t n, const ICompare & compare, v
 
 //=========================================================================
 
+//These constants are probably architecture and number of core dependent
+static const size_t singleThreadedMSortThreshold = 2000;
+static const size_t multiThreadedBlockThreshold = 64;       // must be at least 2!
+
+#ifdef _USE_TBB
+using tbb::task;
+class TbbParallelMergeSorter
+{
+    class SplitTask : public tbb::task
+    {
+    public:
+        SplitTask(task * _next1, task * _next2) : next1(_next1), next2(_next2)
+        {
+        }
+
+        virtual task * execute()
+        {
+            if (next1->decrement_ref_count() == 0)
+                spawn(*next1);
+            if (next2->decrement_ref_count() == 0)
+                return next2;
+            return NULL;
+        }
+    protected:
+        task * next1;
+        task * next2;
+    };
+
+    class BisectTask : public tbb::task
+    {
+    public:
+        BisectTask(TbbParallelMergeSorter & _sorter, void ** _rows, size32_t _n, void ** _temp, unsigned _depth, task * _next)
+        : sorter(_sorter), rows(_rows), n(_n), temp(_temp), depth(_depth), next(_next)
+        {
+        }
+        virtual task * execute()
+        {
+            loop
+            {
+                //On entry next is assumed to be used once by this function
+                if ((n <= multiThreadedBlockThreshold) || (depth >= sorter.singleThreadDepth))
+                {
+                    task * sort = new (next->allocate_child()) SubSortTask(sorter, rows, n, temp, depth);
```

Comment:
The reason is that the continuation for the sort needs to be next.  It is a bit of a pain to do that without creating a task.